onnxruntime/core/mlas/lib/flashattn.cpp

```diff
@@ -0,0 +1,156 @@
+    const MlasFlashAttentionThreadedArgs* args
+)
+{
+    ptrdiff_t row_size_q = static_cast<ptrdiff_t>(args->row_size_q);
+    ptrdiff_t row_size_kv = static_cast<ptrdiff_t>(args->row_size_kv);
+    ptrdiff_t batch_size = static_cast<ptrdiff_t>(args->batch_size);
+    ptrdiff_t num_heads = static_cast<ptrdiff_t>(args->num_heads);
+    ptrdiff_t q_sequence_length = static_cast<ptrdiff_t>(args->q_sequence_length);
+    ptrdiff_t kv_sequence_length = static_cast<ptrdiff_t>(args->kv_sequence_length);
+    ptrdiff_t qk_head_size = static_cast<ptrdiff_t>(args->qk_head_size);
+    ptrdiff_t v_head_size = static_cast<ptrdiff_t>(args->v_head_size);
+    float* buffer = args->buffer;
+    ptrdiff_t buffer_size_per_thread = static_cast<ptrdiff_t>(args->buffer_size_per_thread);
+    ptrdiff_t thread_count = static_cast<ptrdiff_t>(args->thread_count);
+    const float* query = args->query;
+    const float* key = args->key;
+    const float* value = args->value;
+    float* output = args->output;
+
+#if defined(MLAS_TARGET_AMD64) || defined(MLAS_TARGET_LARCH64)
+    auto&& mlas_platform = GetMlasPlatform();
+#endif
+
+    ptrdiff_t q_chunk_count = (q_sequence_length + (row_size_q - 1)) / row_size_q;
+
+    ptrdiff_t task_start = 0;
+    ptrdiff_t task_end = 0;
+    ptrdiff_t total_task_count = batch_size * num_heads * q_chunk_count;
+    ptrdiff_t quotient = total_task_count / thread_count;
+    ptrdiff_t remainder = total_task_count % thread_count;
+    if (thread_id < remainder) {
+        task_start = (quotient + 1) * thread_id;
+        task_end = task_start + quotient + 1;
+    } else {
+        task_start = quotient * thread_id + remainder;
+        task_end = task_start + quotient;
+    }
+
+    for (ptrdiff_t task_index = task_start; task_index < task_end; ++task_index) {
+        ptrdiff_t ib = task_index;
+        ptrdiff_t il = (ib % q_chunk_count) * row_size_q;
+        ib /= q_chunk_count;
+        ptrdiff_t ih = ib % num_heads;
+        ib /= num_heads;
+
+        char* buffer_current_thread = reinterpret_cast<char*>(buffer) + thread_id * buffer_size_per_thread;
+        float* l = reinterpret_cast<float*>(buffer_current_thread);
+        float* m = l + row_size_q;
+        for (ptrdiff_t t = 0; t < row_size_q; ++t) {
+            m[t] = std::numeric_limits<float>::lowest();
+        }
+        float* intermediate = m + row_size_q;
+        float* temp_output = intermediate + row_size_q * row_size_kv;
+        float negmax = 0;
+
+        for (ptrdiff_t ir = 0; ir < kv_sequence_length; ir += row_size_kv) {
+            /*
+                S = Q[ib, ih, il:il+row_size_q, :] * (K[ib, ih, ir:ir+row_size_kv, :]).T
+                old_m = m
+                m = max(m, rowmax(S))
+                diff = old_m - m
+                S = exp(S - m)
+                l = exp(diff) * l + rowsum(S)
+                O = diag(exp(diff)) * O + S * V[ib, ih, ir:ir+row_size_kv, :]
+            */
+            // TODO: Need to concat if past_k is present
```

Comment:
is this TODO still valid? I didn't see `past_k` is defined anywhere.

Reply:
`past_k` is not passed into the kernel but it is an optional input of MultiHeadAttention. If we want to support it in the future, we need to pass it in and concat it with `k`.